File: core/metrics.py
"""
Metrics and readouts, all operate on Shape objects.
The shape subspace is k != 0. C0 is excluded from eta by design (spec §2).
Any readout that claims to measure shape match must exclude C0 explicitly,
or it silently reintroduces the decoupled-term leak (spec §5).
"""

import math

import numpy as np

from .constants import N, GAMMA, ETA_THRESH, EPSILON, k_to_index
from .shape import Shape



# Core metric

def eta(shape_a: Shape, shape_b: Shape) -> float:
    """
    Shape similarity metric (spec §2).

    eta(A, B) = |<A, B>_shape| / (||A||_shape * ||B||_shape)

    Invariant to: global phase rotation, uniform gain, C0 modification.
    Range: [0, 1]. 1.0 = identical shape geometry.
    """
    mask = np.ones(2 * N + 1, dtype=bool)
    mask[k_to_index(0)] = False

    a = shape_a.coeffs[mask]
    b = shape_b.coeffs[mask]

    norm_a = math.sqrt(np.sum(np.abs(a) ** 2))
    norm_b = math.sqrt(np.sum(np.abs(b) ** 2))
    denom = norm_a * norm_b

    if denom < EPSILON:
        return 0.0

    inner = np.sum(a * np.conj(b))
    return float(abs(inner) / denom)



# Readout

def y_rx(shape_in: Shape, target: Shape,
         formula: str = "spec", threshold: float = ETA_THRESH) -> float:
    """
    Scalar readout with match gate (spec §3).

    Args:
        shape_in: received shape S_in.
        target: reference shape T.
        formula:
            "spec"  -> GAMMA * ||S_in||^2_full * eta   (literal spec §3.4)
            "fixed" -> GAMMA * ||S_in||^2_shape * eta  (excludes C0, closes leak §5)
        threshold: eta below this returns 0.0.
    """
    e = eta(shape_in, target)
    if e < threshold:
        return 0.0

    if formula == "spec":
        norm = shape_in.norm_full()
    elif formula == "fixed":
        norm = shape_in.norm_shape()
    else:
        raise ValueError(f"formula must be 'spec' or 'fixed', got {formula}")

    return float(GAMMA * norm * e)



# Energy audit

def energy_audit(before: Shape, after: Shape) -> dict:
    """
    Track energy flow across an operator (spec §3, §4C).

    Returns:
        dict with keys:
            energy_before, energy_after, energy_delta,
            norm_preserved (bool, within EPSILON),
            loss_is_truncation (bool, True if no Class C operator was active)
    """
    e_before = before.norm_full()
    e_after = after.norm_full()
    delta = e_after - e_before

    preserved = abs(delta) < EPSILON

    return {
        "energy_before": e_before,
        "energy_after": e_after,
        "energy_delta": delta,
        "norm_preserved": preserved,
    }

File: socc/src/metrics.c
#include "../include/types.h"
#include "../include/metrics.h"

#include <math.h>
#include <stddef.h>

// eta


double soc_eta(const SocShape* a, const SocShape* b) {
    if (a == NULL || b == NULL) {
        return 0.0;
    }

    double norm_a_sq = 0.0;
    double norm_b_sq = 0.0;
    SocComplex inner = 0.0 + 0.0 * I;

    for (int idx = 0; idx < SOC_COEFF_LENGTH; idx++) {
        if (idx == SOC_N) continue;  // skip k=0

        double mag_a = cabs(a->coeffs[idx]);
        double mag_b = cabs(b->coeffs[idx]);
        norm_a_sq += mag_a * mag_a;
        norm_b_sq += mag_b * mag_b;

        // inner = sum a_k * conj(b_k)
        inner += a->coeffs[idx] * conj(b->coeffs[idx]);
    }

    double norm_a = sqrt(norm_a_sq);
    double norm_b = sqrt(norm_b_sq);
    double denom = norm_a * norm_b;

    if (denom < 1e-12) {
        return 0.0;
    }

    return cabs(inner) / denom;
}


// y_rx


double soc_y_rx(const SocShape* received, const SocShape* target,
                int formula, double threshold, double gamma) {
    if (received == NULL || target == NULL) {
        return 0.0;
    }
    if (formula != SOC_YRX_FORMULA_SPEC && formula != SOC_YRX_FORMULA_FIXED) {
        return 0.0;
    }

    double e = soc_eta(received, target);
    if (e < threshold) {
        return 0.0;
    }

    double norm_sq;
    if (formula == SOC_YRX_FORMULA_SPEC) {
        // Full norm, includes C0 — the §5 leak
        norm_sq = 0.0;
        for (int i = 0; i < SOC_COEFF_LENGTH; i++) {
            double mag = cabs(received->coeffs[i]);
            norm_sq += mag * mag;
        }
    } else {
        // Shape norm, excludes C0 — the fix
        norm_sq = 0.0;
        for (int i = 0; i < SOC_COEFF_LENGTH; i++) {
            if (i == SOC_N) continue;
            double mag = cabs(received->coeffs[i]);
            norm_sq += mag * mag;
        }
    }

    return gamma * norm_sq * e;
}


// energy_audit


SocStatus soc_energy_audit(const SocShape* before, const SocShape* after,
                           SocEnergyAudit* audit) {
    if (before == NULL || after == NULL || audit == NULL) {
        return SOC_ERR_NULL_POINTER;
    }

    double e_before = 0.0;
    double e_after = 0.0;

    for (int i = 0; i < SOC_COEFF_LENGTH; i++) {
        double mag_b = cabs(before->coeffs[i]);
        double mag_a = cabs(after->coeffs[i]);
        e_before += mag_b * mag_b;
        e_after += mag_a * mag_a;
    }

    audit->energy_before = e_before;
    audit->energy_after = e_after;
    audit->energy_delta = e_after - e_before;
    audit->norm_preserved = (fabs(audit->energy_delta) < 1e-12) ? 1 : 0;

    return SOC_OK;
}
